hedera-node/src/main/java/com/hedera/services/ledger/BalanceChange.java

```diff
@@ -0,0 +1,143 @@
+import com.hederahashgraph.api.proto.java.TokenID;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INSUFFICIENT_ACCOUNT_BALANCE;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INSUFFICIENT_TOKEN_BALANCE;
+
+/**
+ * Process object that encapsulates a balance change, either ℏ or token unit.
+ *
+ * Includes an optional override for the {@link ResponseCodeEnum} to be used
+ * in the case that the change is determined to result to an insufficient balance;
+ * and a field to contain the new balance that will result from the change.
+ * (This field is helpful to simplify work done in {@link HederaLedger}.)
+ *
+ * The {@code explicitTokenId} and {@code explicitAccountId} fields are
+ * temporary, needed to interact with the {@link com.hedera.services.ledger.accounts.BackingAccounts}
+ * and {@link com.hedera.services.ledger.accounts.BackingTokenRels} components
+ * whose APIs still use gRPC types.
+ */
+public class BalanceChange {
+	private final Id token;
+	private final Id account;
+	private final long units;
+	private ResponseCodeEnum codeForInsufficientBalance = INSUFFICIENT_ACCOUNT_BALANCE;
+
+	private long newBalance;
+	private TokenID explicitTokenId = null;
+	private AccountID explicitAccountId = null;
+
+	private BalanceChange(Id token, Id account, long units) {
+		this.token = token;
+		this.account = account;
+		this.units = units;
+	}
+
+	public static BalanceChange hbarAdjust(Id account, long units) {
+		return new BalanceChange(null, account, units);
+	}
+
+	public static BalanceChange tokenAdjust(Id token, Id account, long units) {
+		final var tokenChange = new BalanceChange(token, account, units);
+		tokenChange.setCodeForInsufficientBalance(INSUFFICIENT_TOKEN_BALANCE);
+		return tokenChange;
+	}
+
+	public boolean isForHbar() {
+		return token == null;
+	}
+
+	public Id token() {
+		return token;
+	}
+
+	public Id account() {
+		return account;
+	}
+
+	public long units() {
+		return units;
+	}
+
+	public long getNewBalance() {
+		return newBalance;
+	}
+
+	public void setNewBalance(long newBalance) {
+		this.newBalance = newBalance;
+	}
+
+	public TokenID tokenId() {
+		return (explicitTokenId != null) ? explicitTokenId : token.asGrpcToken();
+	}
+
+	public AccountID accountId() {
+		return (explicitAccountId != null) ? explicitAccountId : account.asGrpcAccount();
+	}
+
+	public void setExplicitTokenId(TokenID explicitTokenId) {
+		this.explicitTokenId = explicitTokenId;
+	}
+
+	public void setExplicitAccountId(AccountID explicitAccountId) {
+		this.explicitAccountId = explicitAccountId;
+	}
+
+	public ResponseCodeEnum codeForInsufficientBalance() {
+		return codeForInsufficientBalance;
+	}
+
+	public void setCodeForInsufficientBalance(ResponseCodeEnum codeForInsufficientBalance) {
+		this.codeForInsufficientBalance = codeForInsufficientBalance;
+	}
```

Comment:
I don't see any benefit of having this public setter. We should modify the private constructor to have another param for `codeForInsufficientBalance` and set it accordingly for `hbarAdjust` or `tokenAdjust`.